contracts/RestrictedNORI.sol

```diff
@@ -0,0 +1,1034 @@
+   * https://docs.openzeppelin.com/contracts/3.x/api/token/erc1155#IERC1155-safeBatchTransferFrom-address-address-uint256---uint256---bytes-)
+   */
+  function safeBatchTransferFrom(
+    address from,
+    address to,
+    uint256[] memory ids,
+    uint256[] memory amounts,
+    bytes memory data
+  ) public override {
+    super.safeBatchTransferFrom(from, to, ids, amounts, data);
+    for (uint256 i = 0; i < ids.length; i++) {
+      Schedule storage schedule = _scheduleIdToScheduleStruct[ids[i]];
+      if (amounts[i] != 0) {
+        // slither-disable-next-line unused-return id may already be in set and that is ok
+        _addressToScheduleIdSet[to].add(ids[i]);
+        // slither-disable-next-line unused-return address may already be in set and that is ok
+        schedule.tokenHolders.add(to);
+      }
+      if (balanceOf(from, ids[i]) == 0) {
+        // slither-disable-next-line unused-return return value irrelevant, id guaranteed removed
+        _addressToScheduleIdSet[from].remove(ids[i]);
+        // slither-disable-next-line unused-return return value irrelevant, address guaranteed removed
+        schedule.tokenHolders.remove(from);
+      }
+    }
+  }
+
+  /**
+   * For each index-matched account, removal id, and amount, revokes amount of tokens from schedule corresponding
+   * to the specified removal and transfers to the corresponding account.
+   *
+   * @dev Transfers any unreleased tokens in the removal id's corresponding schedule and reduces the total supply
+   * of that token. No change is made to balances that have released but not yet been claimed.
+   * If a token has multiple owners, balances are burned proportionally to ownership percentage, summing to the total
+   * amount being revoked.
+   *
+   * The behavior of this function can be used in two specific ways:
+   * - To revoke a specific number of tokens as specified by the `amounts` array.
+   * - To revoke all remaining revokable tokens in a schedule set amount to 0 in the `amounts` array.
+   *
+   * ##### Requirements:
+   *
+   * - Can only be used when the caller has the `TOKEN_REVOKER_ROLE`
+   * - The requirements of _beforeTokenTransfer apply to this function
+   * - toAccounts.length == removalIds.length == amounts.length
+   */
+  function batchRevokeUnreleasedTokenAmounts(
+    address[] calldata toAccounts,
+    uint256[] calldata removalIds,
+    uint256[] calldata amounts
+  ) external whenNotPaused onlyRole(TOKEN_REVOKER_ROLE) {
+    if (!(toAccounts.length == removalIds.length)) {
+      revert ArrayLengthMismatch({
+        array1Name: "fromAccounts",
+        array2Name: "removalIds"
+      });
+    }
+    if (!(toAccounts.length == amounts.length)) {
+      revert ArrayLengthMismatch({
+        array1Name: "fromAccounts",
+        array2Name: "amounts"
+      });
+    }
+    for (uint256 i = 0; i < toAccounts.length; i++) {
+      _revokeUnreleasedTokens(toAccounts[i], removalIds[i], amounts[i]);
```

Comment:
seems a little nasty that this internally could load the following several times if some of the removals in the array share a schedule

```
uint256 scheduleId = _removal.getProjectIdForRemoval(removalId);
Schedule storage schedule = _scheduleIdToScheduleStruct[scheduleId];

Removal.ScheduleData memory scheduleData = _removal
      .getScheduleDataForProjectId(projectId);
```

Wonder if this function should restrict removalIds to all have the same schedule

Reply:
interesting... if we did that, we could also simplify a lot of the other logic in this function and get rid of the private implementation inside a loop entirely because it really would just be for a single schedule.  the send would be a single case too, and we wouldn't need an array of `toAccounts`.  We could even consider just passing the `projectId` instead of an array of removal ids, since a revocation really happens at the level of a single schedule/project.  We may still want the removal ids if we choose to add functionality to this function that marks those removals as released at the same time as revocation of any revocable tokens though.  There's a lot we haven't thought through about the end-to-end admin process of handling the situation of invalidated removals.  I think because there's so much unknown here still, I'm going to leave a todo on this function with these thoughts and leave it for a followup PR